### PR TITLE
Developer scripts: Release script

### PR DESCRIPTION
**Pull request by levitte**
The old release script that exists in another repository has aged, and
risks becoming messy beyond maintainability if it's made to deal with
multiple OpenSSL version schemes.

A solution, which has been seen in other projects, is to have the
release script as part of the versioned source tree, and ensure it's
adapted for the ongoing version scheme in that source tree.

This introduces dev/, a directory of OpenSSL developer "stuff".  We
may expand it with other practical scripts to easy development setup
and other similar things that developers may need.  For now, it's the
release script dev/release.sh, with auxilliary files in dev/release-aux/.
The script is self describing, the manual is available by running the
command `./dev/release.sh --manual`.

The dev/ directory shall never appear in a source distribution.

##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Comment by levitte** [comment]
This is in draft because it's not entirely complete yet.  It doesn't upload anything for the moment, and lacks a few environment checks that the old script has.  However, other than that, it should go through the motions correctly.

@mattcaswell, please run some tests.  There should be no danger to anything you do, as this script clones whatever workspace you start from into a subdirectory, and reverting only involves removing that directory and its contents.

File: dev/release.sh [hunk 1]
```diff
@@ -0,0 +1,614 @@
...
+. ./dev/release-aux/release-version-fn.sh
+. ./dev/release-aux/release-state-fn.sh
+
+# Make sure it's a branch we recognise
+current_branch=$(git rev-parse --abbrev-ref HEAD)
+if (echo "current_branch" \
+        | grep -E -q \
+               -e '^master$' \
+               -e '^OpenSSL_[0-9]+_[0-9]+_[0-9]+[a-z]*-stable$' \
+               -e '^openssl_[0-9]+\.[0-9]+\.x$'); then
```
Comment (mattcaswell) [inline dev/release.sh:185]: Why .x?

File: dev/release.sh [hunk 2]
```diff
@@ -0,0 +1,614 @@
...
+fi
+
+# Initialize #########################################################
+
+echo "== Initializing work tree"
+
+# Generate a release branch name.  We also use this as part of the cloned
+# directory name
+new_branch=$(eval $(git cat-file blob HEAD:VERSION)
+             echo "openssl-$MAJOR.$MINOR.x")
```
Comment (mattcaswell) [inline dev/release.sh:202]: .x?

File: dev/release.sh [hunk 3]
```diff
@@ -0,0 +1,614 @@
...
+set_version
+
+if [ -n "$PRE_RELEASE_TAG" ]; then
+    release=$MAJOR.$MINOR.$PATCH-$PRE_RELEASE_TAG$BUILD_METADATA
+    announce_template=openssl-announce-pre-release.tmpl
+else
+    release=$MAJOR.$MINOR.$PATCH$BUILD_METADATA
+    announce_template=openssl-announce-release.tmpl
+fi
+tag="openssl-$release"
```
Comment (mattcaswell) [inline dev/release.sh:286]: Why the change in tag format? Historically we've used a format like "OpenSSL_1_1_1c"

File: dev/release.sh [hunk 4]
```diff
@@ -0,0 +1,614 @@
...
+B<--verbose> |
+B<--debug> |
+B<--help> |
+B<--manual> |
+]
+
+=head1 DESCRIPTION
+
+B<release.sh> creates an OpenSSL release, given current worktree conditions.
+It will refuses to work unless the current branch is C<master> or a release
```
Comment (mattcaswell) [inline dev/release.sh:458]: s/refuses/refuse/

File: dev/release.sh [hunk 5]
```diff
@@ -0,0 +1,614 @@
...
+B<--alpha> is only acceptable if the I<PATCH> version number is zero and
+the current state is "in development" or that alpha releases are ongoing.
+
+B<--beta> is only acceptable if the I<PATCH> version number is zero and
+that alpha or beta releases are ongoing.
+
+=item B<--next-beta>
+
+Use together with B<--alpha> to switch to beta releases after the current
+release is done.
```
Comment (mattcaswell) [inline dev/release.sh:495]: What if we later decide to do another alpha?

File: dev/release.sh [hunk 6]
```diff
@@ -0,0 +1,614 @@
...
+
+=item B<--branch>
+
+Create a branch specific for the I<MAJOR>.I<MINOR>.x release series, if it
+doesn't already exist, and switch to it.  The exact branch name will be
+C<< openssl-I<MAJOR>.I<MINOR>.x >>.
+
+=item B<--no-upload>
+
+Don't upload the produce files.
```
Comment (mattcaswell) [inline dev/release.sh:512]: s/produce/produced/

File: dev/release.sh [hunk 7]
```diff
@@ -0,0 +1,614 @@
...
+Prior to OpenSSL 3.0, the release branches were named
+C<< OpenSSL_I<version>-stable >>, and the release tags were named
+C<< OpenSSL_I<version> >> for regular releases, or
+C<< OpenSSL_I<version>-preI<n> >> for pre-releases.
+
+From OpenSSL 3.0 ongoing, the release branches are named
+C<< openssl-I<MAJOR>.I<MINOR>.x >>, and the release tags are named
+C<< openssl-I<MAJOR>.I<MINOR>.I<PATCH> >> for regular releases, or
+C<< openssl-I<MAJOR>.I<MINOR>.I<PATCH>-alphaI<n> >> for alpha releases
+and C<< openssl-I<MAJOR>.I<MINOR>.<PATCH>-betaI<n> >> for beta releases.
```
Comment (mattcaswell) [inline dev/release.sh:552]: Why the change of" OpenSSL_" to "openssl-"?

File: dev/release.sh [hunk 8]
```diff
@@ -0,0 +1,614 @@
...
+    git checkout $git_quiet -b "$new_branch"
+fi
+
+echo "== Configuring OpenSSL for update and release.  This make take a bit of time"
+
+./Configure cc | while read L; do $VERBOSE "> $L"; done
+
+$VERBOSE "== Checking source file updates"
+
+make update | while read L; do $VERBOSE "> $L"; done
```
Comment (mattcaswell) [inline dev/release.sh:268]: Should we abort if Configure or make fail?

File: dev/release.sh [hunk 9]
```diff
@@ -0,0 +1,614 @@
...
+--no-update     Don't perform 'make update'.
+--verbose       Verbose output.
+--debug         Include debug output.  Implies --no-upload.
+
+--force         Force execution
+
+--help          This text
+--manual        The manual
+
+If none of --major, --minor, --alpha, --beta, or --final are given, this script
```
[the remark this answers is not in the document]
Comment (levitte) [reply dev/release.sh:34]: Thanks, they will disappear

File: dev/release-aux/release-state-fn.sh [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+#               has been created.
+#
+# This function expects there to be a function called fixup_version(), which
+# SHOULD take the |next| as first argument, and SHOULD increase the label
+# counter or the PATCH number accordingly, but only when the current
+# state is "in development".
+
+next_release_state () {
+    local next="$1"
+    local today=$(date '+%-d %b %Y')
```
[the remark this answers is not in the document]
Comment (levitte) [reply dev/release-aux/release-state-fn.sh:47]: Ah.  What shell did you use?  I might have inadvertently used a bash-ism

File: dev/release.sh [hunk 11]
```diff
@@ -0,0 +1,614 @@
...
+while true; do
+    case $1 in
+        --alpha | --beta | --final )
+            next_method=$(echo "x$1" | sed -e 's|^x--||')
+            if [ -z "$next_method2" ]; then
+                next_method2=$next_method
+            fi
+            shift
+            if [ "$next_method" = 'final' ]; then
+                do_branch=true
```
[the remark this answers is not in the document]
Comment (levitte) [reply dev/release.sh:75]: Lines around 206.  A warning, but otherwise inconsequential.  So what I hear is that maybe I should use a different variable name for the implied flags...

File: dev/release.sh [hunk 12]
```diff
@@ -0,0 +1,614 @@
...
+    exit 1
+fi
+
+# Load version functions
+. ./dev/release-aux/release-version-fn.sh
+. ./dev/release-aux/release-state-fn.sh
+
+# Make sure it's a branch we recognise
+current_branch=$(git rev-parse --abbrev-ref HEAD)
+if (echo "current_branch" \
```
[the remark this answers is not in the document]
Comment (levitte) [reply dev/release.sh:181]: Yes it should!

File: dev/release.sh [hunk 13]
```diff
@@ -0,0 +1,614 @@
...
+        | grep -E -q \
+               -e '^master$' \
+               -e '^OpenSSL_[0-9]+_[0-9]+_[0-9]+[a-z]*-stable$' \
```
[the remark this answers is not in the document]
Comment (levitte) [reply dev/release.sh:184]: I made these scripts in such a way that it should be easy to port to 1.1.1 in case there's an interest.  The only things that *should* need a rewrite is `dev/release-aux/release-version-fn.sh`

[the remark this answers is not in the document]
Comment (levitte) [reply dev/release.sh:184]: ... so while I haven't yet planned *to* backport this, I've planned *for the possibility*

File: dev/release.sh (hunk 1 above)
Comment (levitte) [reply dev/release.sh:185]: It's a branch that's going to have all the PATCH releases, just like we have letter releases in pre-3.0.  `x` represents the series of PATCH numbers, and distinguishes the *branch* from releases that would contain numbers only.

I was inspired by other projects I've been involved in which do exactly this.  https://github.com/darktable-org/darktable is one example.

File: dev/release.sh (hunk 3 above)
Comment (levitte) [reply dev/release.sh:286]: The old tag format is derived from what CVS allowed back in the days.  CVS branches and tags were quite limited, [they only allow letters, digits, dashes and underscores](http://web.mit.edu/gnu/doc/html/cvs_8.html).  Git branches and tags, on the other hand, allow periods, and actually most characters (only a few, like `:`, are disallowed, for syntactic reasons).  So there's really no reason to continue with the period-to-underscore conversion.

Also, I see a benefit in having the same name as the tar file.  Copy-n-paste, just add `.tar.gz` and voilà!

File: dev/release.sh (hunk 5 above)
Comment (levitte) [reply dev/release.sh:495]: Tough...  I actually think we need to be strict with ourselves.  But either way, it's also still perfectly possible to simply edit `VERSION` and commit it, with good justification.

File: dev/release.sh (hunk 7 above)
Comment (levitte) [reply dev/release.sh:552]: Like I said in https://github.com/openssl/openssl/pull/11516/files#r406868833, having tags match the tar file could be beneficial.

File: dev/release-aux/openssl-announce-pre-release.tmpl [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+
+   OpenSSL $release has now been made available.  For details of
+   changes and known issues see the release notes at:
+
+        https://www.openssl.org/news/openssl-$MAJOR.$MINOR-notes.html
+
+   Note: This OpenSSL pre-release has been provided for testing ONLY.
+   It should NOT be used for security critical purposes.
+
+   The $LABEL release is available for download via HTTP and FTP from the
```
[the remark this answers is not in the document]
Comment (levitte) [reply dev/release-aux/openssl-announce-pre-release.tmpl:18]: Hah!  Good catch...  this text is an exact copy of what we have in the old script (apart from some variable names), and we never noticed...

File: dev/release-aux/openssl-announce-pre-release.tmpl [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+
+     * https://www.openssl.org/source/
+     * ftp://ftp.openssl.org/source/
+
+   The distribution file name is:
+
+    o $tarfile
+      Size: $length
+      SHA1 checksum: $sha1hash
+      SHA256 checksum: $sha256hash
```
[the remark this answers is not in the document]
Comment (levitte) [reply dev/release-aux/openssl-announce-pre-release.tmpl:30]: I'll have a closer look at the space thing.  There is a hack that detects series of `=` and tries to make it the same length as the previous line, so I guess it's a bug there.

File: dev/release-aux/release-state-fn.sh (hunk 10 above)
[the remark this answers is not in the document]
Comment (levitte) [reply dev/release-aux/release-state-fn.sh:47]: Ooooohhhh!  No, that wasn't it.  I have to surround the `$(date ... )` part with quotes.  What that line gave you today was basically this:

```
    local today=10 Jan 2020
```

`Jan` is, of course, a perfectly valid variable name, but `2020` isn't...

File: dev/release.sh (hunk 3 above)
Comment (levitte) [reply dev/release.sh:286]: I wrote a bit more about it here: https://mta.openssl.org/pipermail/openssl-project/2020-April/001931.html

**Comment by levitte** [comment]
```
cipher_tdes_common.o appears 2 times for the product libcrypto
Died at ./Configure line 2344.
make: *** No rule to make target 'update'. Stop.
```

Yes, thanks, so it wasn't just me.  I tried to fix it, but my fix turned out to be wrong...  fix will come up, but probably past this weekend.

File: dev/release.sh (hunk 8 above)
Comment (levitte) [reply dev/release.sh:268]: I suggest you look at the very top line...  The `-e` will ensure that *any* error aborts.

File: dev/release.sh (hunk 9 above)
[the remark this answers is not in the document]
Comment (levitte) [reply dev/release.sh:34]: Done

File: dev/release.sh (hunk 11 above)
[the remark this answers is not in the document]
Comment (levitte) [reply dev/release.sh:75]: Done, I simply added another variable, `warn_branch`, to tell if the warning should be displayed or not.

File: dev/release.sh (hunk 13 above)
[the remark this answers is not in the document]
Comment (levitte) [reply dev/release.sh:184]: If we decide that we won't backport this, then it's easy to remove that line.

File: dev/release.sh (hunk 4 above)
Comment (levitte) [reply dev/release.sh:458]: Fixed

File: dev/release.sh (hunk 6 above)
Comment (levitte) [reply dev/release.sh:512]: Fixed

File: dev/release-aux/release-state-fn.sh [hunk 16]
```diff
@@ -0,0 +1,202 @@
...
+# 'alpha', 'alphadev'
+#               The source is undergoing a series of alpha releases.
+# 'beta', 'betadev'
+#               The source is undergoing a series of beta releases.
+# These states are computed from $PRE_LABEL and $TYPE
+#
+# The possible |next| values are:
+# 'alpha'       The source tree should move to an alpha release state, or
+#               stay there.  This trips the alpha / pre-release counter.
+# 'beta'        The source tree should move to an beta release state, or
```
[the remark this answers is not in the document]
Comment (levitte) [reply dev/release-aux/release-state-fn.sh:26]: Thanks, fixed

File: dev/release.sh (hunk 12 above)
[the remark this answers is not in the document]
Comment (levitte) [reply dev/release.sh:181]: Fixed

File: dev/release-aux/release-state-fn.sh (hunk 10 above)
[the remark this answers is not in the document]
Comment (levitte) [reply dev/release-aux/release-state-fn.sh:47]: Fixed

File: dev/release.sh [hunk 17]
```diff
@@ -0,0 +1,614 @@
...
+# Update the version information.  This won't save anything anywhere, yet,
+# but does check for possible next_method errors before we do bigger work.
+next_release_state "$next_method"
+
+if $do_branch; then
+    $VERBOSE "== Creating a release branch: $new_branch"
+    git checkout $git_quiet -b "$new_branch"
+fi
+
+echo "== Configuring OpenSSL for update and release.  This make take a bit of time"
```
[the remark this answers is not in the document]
Comment (levitte) [reply dev/release.sh:262]: Fixed

File: dev/release-aux/openssl-announce-pre-release.tmpl (hunk 14 above)
[the remark this answers is not in the document]
Comment (levitte) [reply dev/release-aux/openssl-announce-pre-release.tmpl:18]: Fixed, in both places

File: dev/release.sh [hunk 18]
```diff
@@ -0,0 +1,614 @@
+#! /bin/sh -e
+# Copyright 2020 The OpenSSL Project Authors. All Rights Reserved.
+#
+# Licensed under the Apache License 2.0 (the "License").  You may not use
+# this file except in compliance with the License.  You can obtain a copy
+# in the file LICENSE in the source distribution or at
+# https://www.openssl.org/source/license.html
+
+usage () {
+    cat <<EOF
```
Comment (richsalz) [inline dev/release.sh:10]: Don't you want ```\EOF``` or similar to avoid quote-eating?

File: dev/release.sh [hunk 19]
```diff
@@ -0,0 +1,614 @@
...
+              -l 'no-upload,no-update' \
+              -l 'verbose,debug' \
+              -l 'force' \
+              -l 'help,manual' \
+              -n release.sh -- - "$@")
+eval set -- "$TEMP"
+while true; do
+    case $1 in
+        --alpha | --beta | --final )
+            next_method=$(echo "x$1" | sed -e 's|^x--||')
```
Comment (richsalz) [inline dev/release.sh:69]: like C switch/case ... undent the body of the case/esac?

File: dev/release-aux/openssl-announce-pre-release.tmpl (hunk 15 above)
[the remark this answers is not in the document]
Comment (levitte) [reply dev/release-aux/openssl-announce-pre-release.tmpl:30]: Ouch!  `read` reads words, not lines, not even with just one variable...  so any starting or ending spaces are removed.  I've rewritten the title fixing hack in perl.

File: dev/release.sh (hunk 19 above)
Comment (levitte) [reply dev/release.sh:69]: Do you know how to make `sh-mode` in emacs do that?

File: dev/release.sh (hunk 18 above)
Comment (levitte) [reply dev/release.sh:10]: Did you actually check that quotes are eaten?  Here's what `./dev/release.sh --help` displayed to me:

```
: ; ./dev/release.sh --help
Usage: release.sh [ options ... ]

--alpha         Start or increase the "alpha" pre-release tag.
--next-beta     Switch to the "beta" pre-release tag after alpha release.
                It can only be given with --alpha.
--beta          Start or increase the "beta" pre-release tag.
--final         Get out of "alpha" or "beta" and make a final release.
                Implies --branch.

--branch        Create a release branch 'openssl-{major}.{minor}.x',
                where '{major}' and '{minor}' are the major and minor version
                numbers.

--no-upload     Don't upload to upload@dev.openssl.org.
--no-update     Don't perform 'make update'.
--verbose       Verbose output.
--debug         Include debug output.  Implies --no-upload.

--force         Force execution

--help          This text
--manual        The manual

If none of --major, --minor, --alpha, --beta, or --final are given, this script
tries to figure out the next step.
```

File: dev/release.sh (hunk 19 above)
Comment (levitte) [reply dev/release.sh:69]: ...  it turned out to be easy...

``` elisp
(setq sh-indent-for-case-label 0)
(setq sh-indent-for-case-alt '+)
```

Comment (richsalz) [reply dev/release.sh:69]: almost as easy as "!vi" :)

**Comment by richsalz** [comment]
I have to say I'm surprised you're writing this in shell.  I think perl would be shorter and more understandable.  Yes, seriously.

File: dev/release.sh (hunk 18 above)
Comment (richsalz) [reply dev/release.sh:10]: the joys of shells.  good.

**Comment by levitte** [comment]
> I have to say I'm surprised you're writing this in shell. I think perl would be shorter and more understandable. Yes, seriously.

I disagree.  I did consider it, but in this case, I actually found shell clearer.  If nothing else, this:

``` sh
    git commit -m 'something'
```

compared to this:

``` perl
    system("git commit -m 'whatever'");
    die "yada yada" if $?;
```

I guess I'm more diverse than you give me credit for

File: dev/release.sh (hunk 19 above)
Comment (levitte) [reply dev/release.sh]: I have no idea what that means...

**Comment by richsalz** [comment]
> I guess I'm more diverse than you give me credit for

No criticism intended.  I was just surprised.  And your explanation makes sense.

File: dev/release.sh (hunk 19 above)
Comment (richsalz) [reply dev/release.sh]: It was a joke.  Spawn a subshell and run vi :)

File: dev/release.sh (hunk 8 above)
Comment (mattcaswell) [reply dev/release.sh:272]: ...but it didn't

**Comment by levitte** [comment]
> No criticism intended.

Oh, didn't take it as that.  My remark on diversity was fairly light-hearted :wink:

**Comment by levitte** [comment]
I've added the necessary code to upload as well as to specify a non-default signing key.

**Comment by levitte** [comment]
I've added the fixup code for CHANGES.md and NEWS.md, and done some other fixes.  I believe it's ready for a new review, @mattcaswell

File: dev/release-aux/openssl-announce-pre-release.tmpl [hunk 20]
```diff
@@ -0,0 +1,49 @@
+
+   OpenSSL version $release ($LABEL)
```
Comment (mattcaswell) [inline dev/release-aux/openssl-announce-pre-release.tmpl:2]: Does it make sense to show the patch number here? We also end up with "alpha" twice. This ends up as:

> OpenSSL version 3.0.0-alpha1 (alpha)

Wouldn't this be better as:

> OpenSSL version 3.0 (alpha1)

File: dev/release-aux/openssl-announce-pre-release.tmpl [hunk 21]
```diff
@@ -0,0 +1,49 @@
...
+   ===========================================
+
+   OpenSSL - The Open Source toolkit for SSL/TLS
+   https://www.openssl.org/
+
+   OpenSSL $MAJOR.$MINOR.$PATCH is currently in $LABEL.
```
Comment (mattcaswell) [inline dev/release-aux/openssl-announce-pre-release.tmpl:8]: Similarly here. Not sure about showing the patch number. This ends up as:

>  OpenSSL 3.0.0 is currently in alpha.

To my eye, this is better as:

> OpenSSL 3.0 is currently in alpha

File: dev/release-aux/openssl-announce-pre-release.tmpl [hunk 22]
```diff
@@ -0,0 +1,49 @@
...
+   The $LABEL release is available for download via HTTPS and FTP from the
+   following master locations (you can find the various FTP mirrors under
+   https://www.openssl.org/source/mirror.html):
+
+     * https://www.openssl.org/source/
+     * ftp://ftp.openssl.org/source/
+
+   The distribution file name is:
+
+    o $tarfile
```
Comment (mattcaswell) [inline dev/release-aux/openssl-announce-pre-release.tmpl:27]: I'm even unsure whether the patch number belongs in the filename for pre-releases. This comes out as:

> openssl-3.0.0-alpha1.tar.gz

But, is this better as:

> openssl-3.0-alpha1.tar.gz

File: dev/release.sh [hunk 23]
```diff
@@ -0,0 +1,695 @@
...
+        echo >&2 "Internal option error ($next_method, $next_method2)"
+        exit 1
+        ;;
+esac
+
+# Verbosity feed for certain commands
+VERBOSITY_FIFO=/tmp/openssl-$$.fifo
+mkfifo -m 600 $VERBOSITY_FIFO
+( cat $VERBOSITY_FIFO | while read L; do $VERBOSE "> $L"; done ) &
+exec 42>$VERBOSITY_FIFO
```
Comment (mattcaswell) [inline dev/release.sh:178]: I get this error at this point:

````
dev/release.sh: 178: exec: 42: not found
````

File: dev/release-aux/openssl-announce-pre-release.tmpl (hunk 20 above)
Comment (t8m) [reply dev/release-aux/openssl-announce-pre-release.tmpl:2]: But we are not dropping the last version field in the new versioning scheme, just the letter. So `OpenSSL version 3.0.0 (alpha1)`?

File: dev/release-aux/openssl-announce-pre-release.tmpl (hunk 22 above)
Comment (t8m) [reply dev/release-aux/openssl-announce-pre-release.tmpl:27]: I would not recommend doing 3.0 in pre-release and switching to 3.0.0 after that. It would be confusing.

File: dev/release-aux/openssl-announce-pre-release.tmpl (hunk 21 above)
Comment (t8m) [reply dev/release-aux/openssl-announce-pre-release.tmpl:8]: Here it makes sense because this is rather the name of the release stream.

File: dev/release-aux/openssl-announce-pre-release.tmpl (hunk 20 above)
Comment (mattcaswell) [reply dev/release-aux/openssl-announce-pre-release.tmpl:2]: Right - but does the patch number hold any significance here? Throughout the documentation, when referring to the series of versions we refer to 3.0 - never 3.0.0. In much the same way as we say 1.1.1 as shorthand when referring to 1.1.1a, 1.1.1b, 1.1.1c etc. To me having a specific patch number makes no sense here. It's not an alpha release for a specific patch - its an alpha release for the whole 3.0 series.

Comment (levitte) [reply dev/release-aux/openssl-announce-pre-release.tmpl:2]: Curious, btw, the label should be just "alpha"...  However, I don't really see why we must separate it out, so I would rather throw away the parenthesis altogether.

File: dev/release-aux/openssl-announce-pre-release.tmpl (hunk 22 above)
Comment (levitte) [reply dev/release-aux/openssl-announce-pre-release.tmpl:27]: No, but it would make sense to have openssl-3.0.tar.gz as the final release.  The patch releases that follow it would of course be openssl-3.0.1.tar.gz, openssl-3.0.2.tar.gz etc

File: dev/release.sh (hunk 23 above)
Comment (levitte) [reply dev/release.sh:178]: What shell are you using?  It clearly doesn't follow POSIX specs in that case...

Comment (mattcaswell) [reply dev/release.sh:178]: bash

Comment (levitte) [reply dev/release.sh:178]: See https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_07_05

Comment (levitte) [reply dev/release.sh:178]: Also https://pubs.opengroup.org/onlinepubs/009695399/utilities/exec.html

Comment (levitte) [reply dev/release.sh:178]: Er...  I use bash as well, and famous last words, it works for me.

File: dev/release-aux/openssl-announce-pre-release.tmpl (hunk 22 above)
Comment (t8m) [reply dev/release-aux/openssl-announce-pre-release.tmpl:27]: In theory, yes, but it would be still confusing.

File: dev/release-aux/openssl-announce-pre-release.tmpl (hunk 20 above)
Comment (t8m) [reply dev/release-aux/openssl-announce-pre-release.tmpl:2]: It would be really confusing if we released the final version as 3.0 and then the patch levels were 3.0.1 and so on. And yes the 3.0 refers to the whole series but a concrete release should not lose and gain the patch-level number. At least it is very uncommon among the open-source projects to do so.

File: dev/release.sh (hunk 23 above)
Comment (levitte) [reply dev/release.sh:178]: @mattcaswell and I found that /bin/sh is often linked to dash on Debian-based systems, and that dash isn't quite POSIXly correct, so I'm switching the shebang to specify bash.

File: dev/release-aux/openssl-announce-pre-release.tmpl (hunk 22 above)
Comment (levitte) [reply dev/release-aux/openssl-announce-pre-release.tmpl:27]: I'm sorry, what part was confusing now?

File: dev/release-aux/openssl-announce-pre-release.tmpl (hunk 20 above)
Comment (levitte) [reply dev/release-aux/openssl-announce-pre-release.tmpl:2]: I hear you.  I just had a quick look at diverse software version histories, and indeed, those following a semantic version scheme (or semantic version-ish) do present all three numbers at all times, and only use MAJOR.MINOR to represent the series (i.e. 3.0 would be equivalent to 3.0.x).

Comment (levitte) [reply dev/release-aux/openssl-announce-pre-release.tmpl:2]: The benefit with a stricter three-number release version is that it makes the script simpler.

File: dev/release-aux/openssl-announce-pre-release.tmpl (hunk 22 above)
Comment (levitte) [reply dev/release-aux/openssl-announce-pre-release.tmpl:27]: > I'm sorry, what part was confusing now?

Never mind that...

File: dev/release-aux/openssl-announce-pre-release.tmpl (hunk 21 above)
Comment (levitte) [reply dev/release-aux/openssl-announce-pre-release.tmpl:8]: Related to that, @t8m, would you rather the release branch be called `openssl-3.0.x` or would you prefer `openssl-3.0`?  I do prefer the former, but can go with the latter if that's what everyone else prefers.

Comment (t8m) [reply dev/release-aux/openssl-announce-pre-release.tmpl:8]: I do not really have a preference here, both are OK.

Comment (levitte) [reply dev/release-aux/openssl-announce-pre-release.tmpl:8]: I think I'll keep `.x`, as it makes it more distinct

File: dev/release-aux/openssl-announce-pre-release.tmpl (hunk 20 above)
Comment (levitte) [reply dev/release-aux/openssl-announce-pre-release.tmpl]: I've made changes that clarify all this

**Comment by levitte** [comment]
Per a review discussion, the latest fixup makes it more consistent if we talk in terms of a release series or a specific release version.  Alpha and beta releases are always considered to be for whole releases (we never do alphas and betas within a series, so that makes sense).  The master branch is always considered to be the development of the next series (i.e. the next minor or major release).

I believe I've been able to reflect that line of thinking.

**Comment by t8m** [comment]
LGTM, I'll leave formal approval to someone from OMC as this seems to me to be somewhat within their area.

**Comment by levitte** [comment]
> LGTM, I'll leave formal approval to someone from OMC as this seems to me to be somewhat within their area.

I *am* someone from OMC :wink:

**Comment by t8m** [comment]
I know, but I mean the versioning could be a little bit controversial topic so having someone else from OMC other than the author shouldn't hurt.

File: dev/release.sh [hunk 24]
```diff
@@ -0,0 +1,711 @@
...
+. $HERE/dev/release-aux/release-version-fn.sh
+. $HERE/dev/release-aux/release-state-fn.sh
+
+# Make sure it's a branch we recognise
+orig_branch=$(git rev-parse --abbrev-ref HEAD)
+if (echo "$orig_branch" \
+        | grep -E -q \
+               -e '^master$' \
+               -e '^OpenSSL_[0-9]+_[0-9]+_[0-9]+[a-z]*-stable$' \
+               -e '^openssl_[0-9]+\.[0-9]+\.x$'); then
```
Comment (mattcaswell) [inline dev/release.sh:213]: This should be "openssl-" instead of "openssl_" (hyphen not underscore). The tar files are using a hyphen for the version name, and the "--branch" option creates it with a hyphen.

File: dev/release-aux/openssl-announce-pre-release.tmpl [hunk 25]
```diff
@@ -0,0 +1,49 @@
...
+
+   OpenSSL - The Open Source toolkit for SSL/TLS
+   https://www.openssl.org/
+
+   OpenSSL $series is currently in $label.
+
+   OpenSSL $release_text has now been made available.  For details of
+   changes and known issues see the release notes at:
+
+        https://www.openssl.org/news/openssl-$series-notes.html
```
Comment (mattcaswell) [inline dev/release-aux/openssl-announce-pre-release.tmpl:13]: Just to note that this page (perhaps not unsurprisingly) does not yet exist for 3.0. Do we need to update something to make sure that it is?

Comment (levitte) [reply dev/release-aux/openssl-announce-pre-release.tmpl:13]: Yes we do, but that's in openssl/web.

File: dev/release.sh (hunk 24 above)
Comment (levitte) [reply dev/release.sh:213]: Yeah, that was a blunder.  Thanks!

Comment (levitte) [reply dev/release.sh:213]: Fixed

File: dev/release.sh (hunk 23 above)
Comment (kaduk) [reply dev/release.sh:181]: [obligatory note that assuming bash is present is a linux-ism]

Comment (paulidale) [reply dev/release.sh:181]: OS/X has moved from bash to zsh, I wouldn't be entirely surprised if bash disappeared after a couple more releases.

Comment (levitte) [reply dev/release.sh:181]: I try to write POSIXly correctly. If zsh is POSIXly correct, I don't anticipate problems.

**Comment by levitte** [comment]
Have we settled everything here?

File: dev/release-aux/openssl-announce-pre-release.tmpl [hunk 26]
```diff
@@ -0,0 +1,49 @@
...
+   OpenSSL version $release_text released
+   ======================================
+
+   OpenSSL - The Open Source toolkit for SSL/TLS
+   https://www.openssl.org/
+
+   OpenSSL $series is currently in $label.
+
+   OpenSSL $release_text has now been made available.  For details of
+   changes and known issues see the release notes at:
```
Comment (mattcaswell) [inline dev/release-aux/openssl-announce-pre-release.tmpl:11]: Adding a link to the wiki notes would be a useful addition. Either in this PR or a subsequent one.

Comment (levitte) [reply dev/release-aux/openssl-announce-pre-release.tmpl:11]: Other PR

**Comment by levitte** [comment]
Merged.

b0b0b6a41dd4418bab19cdd28c01ce248fee6af7 Developer scripts: Release script
